pkg/remediator/remediator.go

```diff
@@ -175,8 +104,7 @@ func (r *Remediator) Pause() {
 	defer r.lifecycleMux.Unlock()
 
 	klog.V(1).Info("Remediator pausing...")
-	r.stopFn() // tell the workers to stop
-	<-r.doneCh // wait until workers are done (if running)
+	r.canRemediate = false
```

Comment:
This will prevent new remediator executions from starting after `Pause`, but does not guarantee that a prior execution is not still running. This means that if a remediator Reconcile starts before a syncer Reconcile, they could be running concurrently. The syncer still needs to block and wait for remediator executions to complete

Reply:
The remediator reconciles a single object at a time. When a remediator reconcile starts before a syncer reconcile, there are at most one remediator reconciliation. When it finishes, remaining remediations will be skipped and requeued, so there won't have continuous fights.

If we block the Syncer while Remediator is processing each object, there will be very frequent pause and resume signals for the Syncer, which might be an issue.

Reply:
This is fundamentally a race condition which could lead to surprising/unexpected behavior, so should be prevented IMO. The remediator generally should only be running between syncs, and never concurrently with syncer.

It seems essential for the syncer to send a pause signal to the remediator and receive a done signal before proceeding.

Reply:
If the syncer needs to block and wait for remediator to complete, it is possible that the syncer is blocked by remediator errors because those errors will be retried.

Reply:
In this example I think it would be worse to let the syncer run while there's a remediator thread retry-looping, as this just increases the risk of a race condition (remediator thread continuing to run concurrently during syncer execution).

To minimize the impact of this we can wrap the remediator thread in a context with a reasonable timeout for the remediator (I expect this should probably be relatively short for remediator). We could also potentially cancel the context in the Pause function

Reply:
Syncer will pause the remediator. With the current logic, if an object fails to be remediated, it keeps retrying until the next syncer kicks in. The syncer pauses the remediator, so no new remediation will run. After the syncer finishes applying, the remediate queue will be updated, and the remediator is resumed.
When dequeuing the object, it may already get updated, so becomes a no-op. If not, it continues with retry.

Reply:
I was responding to your previous comment, which implies the remediator may continue retry-looping after being paused

> If the syncer needs to block and wait for remediator to complete, it is possible that the syncer is blocked by remediator errors because those errors will be retried.

Reply:
Another caveat inherited from the previous architecture is that the Syncer applies a list of declared objects, while the Remediator reconciles a single object. I'm thinking if we should change the Remediator to process a list of resources. If so, then it becomes a mini-apply. Do we want to use the kpt applier to remediate them, and how do we handle the fight?

Reply:
> I was responding to your previous comment, which implies the remediator may continue retry-looping after being paused
> 
> > If the syncer needs to block and wait for remediator to complete, it is possible that the syncer is blocked by remediator errors because those errors will be retried.

Nope, it won't. After being paused, the Remediator's request will be skipped (but requeued) if `canRemediate` is false.

Reply:
In that case, it sounds like the syncer just needs to wait for the remediator to complete any prior requests after pausing

Reply:
I see. So basically when the syncer applies declared objects, it:
1. pause the Remediator
2. wait for the Remediator to be paused
3. start applying
4. resume the Remediator if apply succeeds.

Reply:
Yes, that's how I see it. If we don't perform step 2 (`wait for the Remediator to be paused`) then we allow the possibility for unpredictable race conditions

Reply:
That seems to go hand-in-hand with your other comment. After waiting for Remediator to be paused, we can clear the remediate queue. The only caveat is to not requeue the request with the controller-runtime library, but only use the remediate queue for requeueing and dequeueing.

Reply:
done